src/cobald/daemon/runners/service.py

```diff
@@ -12,6 +13,19 @@
 from ..debug import NameRepr
 
 
+T = TypeVar("T")
+
+
+def _weakset_copy(ws: "weakref.WeakSet[T]") -> Set[T]:
+    """Thread-safely copy all items from a weakset to a set"""
+    # The various WeakSet methods are not thread-safe because they miss locking.
+    # The main issue is that all copy approaches use ``__iter__``, which is not
+    # thread-safe against items being GC'd. However, we can access the actual
+    # backing real set ``ws.data`` and ``set(some_set)`` is GIL-atomic.
```

Comment:
I've tried to find a proper reference on this but failed. [This SO Q&A](https://stackoverflow.com/questions/52693294/is-it-safe-to-give-a-python-weakset-to-a-list-constructor) outlines that directly copying/iterating ``WeakSet`` itself is *not* threadsafe, but no alternatives. The comment here is the TLDR of me spelunking through the CPython and PyPy 3.6 source code.